ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/StateTreeManager.java

```diff
@@ -83,81 +86,109 @@ public void onValidatorRegistered(ValidatorRegistrationEvent event) {
           this.store.addProcessedBlock(blockStateRoot, block.get());
           this.store.addProcessedBlock(blockRoot, block.get());
-        } else {
-          LOG.info(
-              "Receieved genesis block.  Checking to see if state root matches this clients initial state root");
-          Optional<BeaconState> state = this.store.getState(block.get().getState_root());
-          if (state.isPresent()) {
-            LOG.info("State roots match! Saving genesis block in storage");
-            this.store.addProcessedBlock(blockStateRoot, block.get());
-            this.store.addProcessedBlock(blockRoot, block.get());
-            newState = BeaconState.deepCopy(state.get());
+          this.store.addState(newStateRoot, newState);
+
+          // run stateTransition.initiate() on slots from block.slot to node.slot
+          counter = 0;
+          while (newState.getSlot().compareTo(nodeSlot) < 0) {
+            if (counter == 0) {
+              LOG.info(
+                  "Transitioning state from slot: " + newState.getSlot() + " to slot: " + nodeSlot);
+            }
+            newState = BeaconState.deepCopy(newState);
+            stateTransition.initiate(newState, null);
+            newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
+            this.store.addState(newStateRoot, newState);
+            counter++;
           }
+          LOG.info("latest state root: " + newStateRoot.toHexString());
+          this.canonical_state = newState;
+        } else {
+          LOG.info("The block's state root does not match the calculated state root!");
+          LOG.info("  new state root: " + newStateRoot.toHexString());
+          LOG.info("  block state root: " + blockStateRoot.toHexString());
         }
       } else {
-        // TODO: If there is no block for the current slot, should i just get the canonical state?
         newState = BeaconState.deepCopy(this.canonical_state);
         stateTransition.initiate(newState, null);
+        newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
+        this.store.addState(newStateRoot, newState);
+        LOG.info("latest state root: " + newStateRoot.toHexString());
+        this.canonical_state = newState;
       }
-      this.canonical_state = newState;
-      Bytes32 newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
-      this.store.addState(newStateRoot, newState);
     } catch (NoSuchElementException | IllegalArgumentException | StateTransitionException e) {
       LOG.warn(e);
     }
   }
 
+  protected Boolean shouldProcessBlock(Optional<BeaconBlock> block) {
+    if (!block.isPresent()) {
+      return false;
+    }
+    if (!this.store.getParent(block.get()).isPresent()) {
+      return false;
+    }
+    UnsignedLong blockTime =
+        UnsignedLong.valueOf(block.get().getSlot())
+            .times(UnsignedLong.valueOf(Constants.SLOT_DURATION));
+    if (this.nodeTime.compareTo(blockTime) < 0) {
```

Comment:
currently, if this function returns false due to nodeTime being less than blockTime, we initiate a state transition with a null block. is that what we want to do?